proposals/3/README.md

```diff
@@ -0,0 +1,210 @@
+   </td>
+   <td>
+    <a href="https://github.com/jenkins-x/jx/issues/6734">#6734</a>,
+    <a href="https://github.com/jenkins-x/jx/issues/6763">#6763</a>
+   </td>
+  </tr>
+  <tr>
+   <td><strong>Story / Task</strong>
+   </td>
+   <td>Split cloud resource creation out of jx boot
+   </td>
+  </tr>
+</table>
+
+# 1. Overview
+
+This document outlines the work for extracting cloud resource creation from jx. 
+
+
+## 1.1 Motivation
+
+At the moment cloud resources needed by Jenkins X, eg service accounts, buckets, cryptographic keys, are generated on the fly.
+This happens either as part of a `jx [create|install]` or via `jx boot` and its various step commands.
+This proposal outlines an approach to separate cloud resource creation from the Jenkins X install. 
+
+
+## 1.2 Background
+
+The motivation for this proposal is driven by the following drawbacks of the current approach:
+
+
+
+*   For users, it is not transparent what cloud resource and permissions are needed  by  Jenkins X
+*   It is hard to limit permissions. At the moment Boot requires full admin permissions on all Cloud APIs
+*   Code for generating and dealing cloud resources is distributed across multiple commands and files in the source code
+*   Hard to integrate new cloud provider due to the lack of abstraction
+*   Hard to maintain current code
+    *   One confusion, for example, is that the various “_verify”_ pipeline steps do not purely verify resources, but also lazily creation 
+
+The following design proposes a clearer separation of concerns, addressing the points mentioned above as well as introducing the required abstractions to integrate with further cloud providers.
+
+
+# 2. Design
+
+
+## 2.1 UX
+
+For the user, the proposed change means that he is responsible for creating all required cloud resources up front (prior to installing Jenkins X).
+We provide documentation on what’s required as well as Terraform scripts to create and manage the required resources. 
+
+When using Terraform, `terraform apply` is effectively replacing current `jx create`.
+
+If a Jenkins X feature requires new Cloud resources, upgrading the cluster will fail until the user has created the new required resource, either manually or via the Terraform scripts. 
+
+The Jenkins X install process will always only verify whether a needed resource exists, never create it
+When verification fails, the install process halts and needs to provide enough context to the user to identify which (cloud) resource is missing.
+
+
+## 2.2. Technical Design
+
+
+### 2.2.1 Docs
+
+Ar part of this work we want to clean up and extend the documentation.
+The updated documentation should include sections for the following:
+
+
+
+*   Generic description of the cloud resources needed and their purpose
+*   Documentation on where to find the Terraform script and how to use them
+
+
+### 2.2.2 Terraform
+
+As part of the work, we are creating a GitHub repository with two Terraform setups.
```

Comment:
I wonder if including Terraform in the early work is a little overkill, it might make things a bit more complicated.  Just a thought as I think starting with documenting the different cloud provider commands along with tutorials etc might be enough to start with and then given the feedback prioritise the work needed to include Terraform?  FWIW for GKE we started listing some commands here which we run and works really nicely https://github.com/jenkins-x-labs/jenkins-x-installer#provision - that way we could focus on good cloud provider integration then bring support for Terraform as a next step?

Reply:
I agree - we should have wiki page / docs on the canonical commands for GKE / AWS / AKS (e.g. `gcloud` or `eksctl` or whatever) first before the terraform; as those are known pretty well already and can be used by anyone

Reply:
> I wonder if including Terraform in the early work is a little overkill, it might make things a bit more complicated.

More complicated for whom? For the user? Or for us to develop?

> FWIW for GKE we started listing some commands here which we run and works really nicely https://github.com/jenkins-x-labs/jenkins-x-installer#provision 

I've seen this one. Very useful. The last time I looked the commands were mainly around creating the service accounts and setting up workload identity. We would also need commands to create the buckets and everything around creating a Vault instance (https://github.com/sethvargo/vault-on-gke). 

Documenting everything as commands to be executed using _gcloud_ is an option, but it has pro and cons. An advantage of the terraform script is that it is runnable and partly self-documenting "code". Pages in the documentation have a much higher risk to get out of sync. 

For GKE some of the heavy lifting is already provided by https://github.com/jenkins-x/terraform-google-jx. 

What do others think?

Reply:
> that way we could focus on good cloud provider integration 

Right, key here is the documentation, but I had more of generic documentation in mind which list all the things which are needed. The provided Terraform script is then a "codified" version of that. 

With the docs in place, everyone should be able to manually create the required resources if he wants to. Work-wise we should also be able to decouple the changes to `jx boot` from development of the Terraform scripts (or the 
_gcloud_  / _eksctl_ scripts for that matter, if we decide this is the better approach).

Reply:
I guess my worry would be focussing on Terraform means anyone not already familiar with it then needs to grok that first and increases the barrier to entry.

Reply:
> I wonder if including Terraform in the early work is a little overkill, it might make things a bit more complicated.

> More complicated for whom? For the user? Or for us to develop?

My thinking is both users and developing / supporting it, my vote would be to focus on the cloud provider commands first and then Terraform later if we get lots of people asking for it?

Reply:
My thoughts here are that both terraform and good docs around the resources that need to be created are complementary. If we provide both we make it easy for both types of users. The use of terraform should make it easier to provide a reference architecture for provisioning resources that could be followed by other cloud providers if they wish to provide an implementation. It also makes it easier to keep the infrastructure in sync when we roll out changes.

Also the work that helps enable this (removing the cloud resource creation out of boot) would enable contributors to use the cloud resource provisioning technology of their choice if they do not wish to use terraform.

Reply:
Yeah agree with all that, I'm not saying Terraform isn't useful or a good approach, it's more about priorities and where we start.  My gut feeling is to start with one approach rather than two as it's probably going to be faster to do that and folks can get the benefits, then look to introduce Terraform as a second phase.  If including Terraform isn't much of an overhead then cool doing both sounds ok but I would think there is more engineering effort and support needed to include Terraform?

Reply:
i.e. having great support for GKE, EKS, AKS, PKS, OpenShift etc would have my vote to tackle first rather than adding Terraform.

Reply:
> "My gut feeling is to start with one approach rather than two"

What do you mean with two approaches? The proposal is about using Terraform. That together with some general description/documentation on what's required by Jenkins X. Even if you went down the _gcloud_ commands approach you would need the general documentation, right?

I am also no sure how much faster a pure _gcloud_ script would be. ATM I have a feeling that we have more pieces in places using Terraform opposed to plain script.

Reply:
ah ok so probably confusion on my side, I was expecting both Terraform and vanilla cloud provider commands with docs (possible as a script) to be produced with this proposal.  I would still rather focus on the latter and get more cloud provider coverage with BDD tests validating changes on everything we support, IMO rather than Terraform TBH.  Having said that if it's decided against my recommendation then I would consider a few more things:

- UX for developers, i.e what is the flow that a developer will take to create a cluster on a cloud provider, can it be as simple and painless as possible without prior knowledge of Terraform, plus is the UX nice a slick.  So an idea of steps on what the install docs would look like would be great.
- I don't know Terraform particularly well but who would run it and where does the state get stored?  Would we roll out updates? Would folks fork the Terraform repo or use the Jenkins X master?
- Are we saying that to create a cluster for Jenkins X to run on people should use Terraform?
- Maintainability: will there be an overhead supporting Terraform as the primary approach to creating cloud resources? 
- Can Terraform generate the exact vanilla cloud provider commands that would be used so we can have this idea of living documentation?

Reply:
> UX for developers, i.e what is the flow that a developer will take to create a cluster on a cloud provider, can it be as simple and painless as possible without prior knowledge of Terraform, plus is the UX nice a slick. So an idea of steps on what the install docs would look like would be great.

Something along these lines:

```bash
$ git clone jx-terraform
$ cd jx-terraform/gke
$ terraform init
$ terraform apply
```

> I don't know Terraform particularly well but who would run it and where does the state get stored? Would we roll out updates? Would folks fork the Terraform repo or use the Jenkins X master?

Users would clone/fork the Terraform repo we provide. What we do with the state file needs to be discussed. Either just recommendation on the docs or we setup per default to store the state in a bucket. The who is the admin. This might be a single person or multiple.

> Would we roll out updates

This one touches on a good point. How to ensure that the Terraform script you are running is in sync with the Jenkins X version you want to install. In the beginning, this might be just based on documentation. In the long run the Terraform script could add the actual Jenkins X install step at which point the version is set. Might be worth adding something around this to the proposal


Are we saying that to create a cluster for Jenkins X to run on people should use Terraform?
Maintainability: will there be an overhead supporting Terraform as the primary approach to creating cloud resources?
Can Terraform generate the exact vanilla cloud provider commands that would be used so we can have this idea of living documentation?

> Are we saying that to create a cluster for Jenkins X to run on people should use Terraform?

That would be the recommendation yes.

> Maintainability: will there be an overhead supporting Terraform as the primary approach to creating cloud resources?

In which form? I don't think this is more overhead than maintaining _gcloud_ scripts. If we want to keep both, there is a bit of overhead. 

> Can Terraform generate the exact vanilla cloud provider commands that would be used so we can have this idea of living documentation?

That's a good question. TBH, I don't know. Worth investigating.

Reply:
I think there are two dimensions here - the ease of building the solution, and the ease of use by users. 

Focusing first on building Terraform has a (small) upfront cost (mostly because there isn't always great examples around), however has huge maintainence benefits (e.g. can easily validate drift to ensure that everything needed is captured in the script, record of changes, ability to migrate to newer versions etc.). To me, terraform is clearly advantageous over a script for us, and well worth any upfront cost.

When it comes to UX, I think both approaches are equal for users who want the default setup, or those who want to tweak parameters (e.g. autoscaling on/off). I think the big difference comes with larger customizations, where I think almost all of our target audience will understand scripts vs a smaller number of will understand terraform. Overall this doesn't worry me as I think that if you want to make large customizations then taking the time to learn terraform will be acceptable - particularly as terraform isn't that hard to work with.

Reply:
Because Terraform provides breadth across major cloud providers it is good to pilot the effort with Terraform samples first.  The primary concern would be to ensure we deeply describe exactly what the resources that are needed are.  Having multiple approaches to resource creation is definitely advantageous in the long run.

Reply:
I think we are overestimating the amount of effort someone would need in order to understand Terraform.

There are actually two ways of viewing this:

* **Someone who wants to try Jenkins X**: The UX for this persona would be to either run `terraform apply` with a variables file and get everything created for you and a `jx-requirements.yml` file generated or go through a bunch of `gcloud`/ `awsCLI` commands to get everything set. 

The knowledge of Terraform here would need to be minimum, we'd need to ensure that the happy path works and that is executing a single Terraform Script.

* **Someone who wants to MODIFY the Terraform Script**: This is the advanced case, maybe they don't want to create a VPC because they already have one? Maybe they don't want to create buckets because they already have a few they can use? Then assuming some knowledge on Terraform is OK here, because this is an advanced use case and we are still facilitating the work.

In my opinion, focusing on Terraform first while we figure out a way to document every single CLI call that an user would have to do if they don't want to use Terraform is the right way to go.

We don't have to delete the creation of these resources from the source code straight away, and that will help us get users moved to the new mechanism.

Reply:
my worry of going all in on terraform and not showing folks how they could use, say, gcloud on GKE or eksctl/aws CLI on AWS is folks will not even bother trying Jenkins X if a team doesn't already use and know terraform.

If your company already uses another tool (cloud formation / shell scripts / ansible / whatever) and we show folks terraform I suspect they'll just turn away and not go any further.

Maintaining docs [like these](https://github.com/jenkins-x-labs/jenkins-x-installer#create-a-new-cluster-on-gcp-with-googles-workload-identity-enabled) of how you _could_ use just gcloud / AWS CLI means anyone using another tool can easily figure out how to create the cloud resources. Its also very easy to BDD test those CLI instructions.

I agree - if we can get terraform working well and if users want to use terraform - its a good solution. My worry is not everyone agrees on terraform

Reply:
I wouldn't say this proposal is going all in on terraform so much as _starting_ with terraform - it makes sense to start with something that works with both of the fully supported providers, and then follow up with provider-specific alternatives. The initial scope here is "separate out the cloud resource creation", and Terraform is the best first implementation for us to do that, since we need to make sure that we can do the separation in a smooth, user-friendly way for both of the fully supported providers before we can be sure this is actually going to be viable. Adding gcloud or eksctl/aws equivalents quickly after we've proven that removing this stuff from boot's scope is viable and usable seems like the right approach to me.

Reply:
i.e., the important thing is to remember that the Terraform scripting/documentation isn't the end state, it's a step towards the end state.

Reply:
Just my two cents - I don't see anything here saying that we won't / are against documenting the cloud provider commands as an alternative to Terraform eventually, just that Terraform is the best end to end experience we can provide and we should make that really good and button it up and then we can expand approaches later. Does anyone disagree with that?

Assuming that's the case, I think starting with Terraform makes sense for all the experience benefits it provides - and we are likely overestimating how intimidating it is. Jenkins X bundles in / uses quite a few things that are more unknown to most people than Terraform is, so I don't see Terraform as being a big breaking point for that many people (and it's not hard to use.) But that's also not a reason to not document the cloud commands eventually.

Reply:
I still believe starting with simple scripts + good docs that cover cloud solutions (including OpenShift which most wont use Terraform) would have been a better starting point.  We already had a lot done for GKE in the [installer chart](https://github.com/jenkins-x-labs/jenkins-x-installer#harder-way) which took half a day to produce.  Having said that if everyone else believes Terraform is the right starting point and there's not much overhead plus we wont negatively impact UX or the getting started experience then ok but I would like early feedback from the Jenkins X community + users asap and let them feed into UX.

I think there's two main areas to be investigated from the comments above:
1. how to handle terraform state
2. how to handle updates and roll them out

Reply:
@rawlingsj We can also work on coming up with a feedback survey for the community to roll out alongside the change and proactively seek input.

Reply:
> We can also work on coming up with a feedback survey for the community to roll out alongside the change and proactively seek input

@ethangj big +1, sounds great

Reply:
I like the idea of a feedback survey as well. It is a lot of speculation here on what's easier for the user and us as well as what feels intimidating for a user and what not. Doing a survey once we have something implemented seems like a good way to make decisions on actual data. 

We need to start with something. ATM most in this thread are fine with Terraform first. I think this is what we should aim for first and it does not mean that this is the final destination.

Reply:
> I think there's two main areas to be investigated from the comments above:

> how to handle terraform state

Right. We could just go for local state initially. This makes it easier to get started. As an advanced user you might have your own preferred choice of remote backend. 

The alternative is to include already per default a remote backend, but it might be overkill for now. 

Definitely needs some more investigation.

On that note, terraform state handling might be a topic to be investigated, but the state itself is imo a win compated to the pure _gcloud_ commands where one does not have state at all which makes it harder to keep the current state as well as deleting the resources. 

> how to handle updates and roll them out

I guess the problem is not so much about updates to the repo per se, but the question of how you sync a terraform script version with a jx version. For now, that would be a documentation issue imo. 

That said, does the use of _gcloud_ script not have the same problem? If these scripts live in a different repo or the online docs, what ensures that they are in sync with a given _jx_ release. I guess we are in the same boat there.

@rawlingsj did you have something in mind to solve the "update" issue for the script approach?

Reply:
Something I am also missing in this discussion is that the _gcloud_ scripts in the installer repo seem simple, but imo they don't look terribly resilient. What happens if an error occurs halfway through the script?

Reply:
I am wondering what @cagiti or @garethjevans with their Terraform experience have to say to this?

Reply:
My thoughts on this one are that one of core objectives of this proposal is to decouple cloud resource creation from the main install mechanism of Jenkins X (Boot). Excellent documentation around the install dependencies for Jenkins X is a key component of this proposal and is a requirement for it to be delivered successfully. 

Providing a really good out of the box experience here is really important and the use of terraform provides us with a several benefits:

- Hides away the complexity of creating infrastructure 
- Provides a code first approach to creating infrastructure for users that do want to get in to the infrastructure side of Jenkins X
- Adds the ability for a customer to utilise gitops for the infrastructure if they require it
- Adds the ability to add updates to infrastructure as we roll out new features 
- Delegates multiple cloud provider support is to a tool that specialises in it
- Provides a reference for integrating additional cloud providers

But terraform should not be mandatory for users to use Jenkins X. The documentation should contain what is necessary for a user to be able to get up and running with Jenkins X and delegate the cloud resource creation to what is standard practice within their company.

For new users that currently do not have a defined best practice within the company to be guided towards using playbooks for managing infrastructure is a good thing in my opinion as its considered best practice within the industry. Providing something for users to get started with is a nice nudge in a good direction. They can also be used as a reference for integration in to other technologies that can be used to create infrastructure playbooks.

From a documentation point of view the main discussion here seems to be around if the documentation should include cloud provider specific commands that users could use to create the infrastructure. I think this would be a good thing to include in the documentation if we can find a way of producing and keeping up to date that documentation for all cloud providers we support. My initial thinking here was that it would be nice if terraform could help produce cloud provider specific commands, unfortunately on deeper investigation it seems that is not currently a simple thing to do as terraform uses cloud provider APIs over calling out to command line tools. It would though be good to further investigate how much we could help users that do not wish to use terraform and what the documentation structure would be that would enable the most customer success here.

Reply:
I don't think that anyone wanting to run JX in production should be using the installer approach.  It's great for demos/trials but unless it can be idempotent and handle updates, I would expect anyone to use it.

Whilst terraform is a great approach to this, its not a silver bullet.  State management needs to be thought about carefully, for the terraform-google-jx module, we have deliberately not included any state management so it can hook into what ever system a user prefers - e.g. it integrates seamlessly with cloudbees/jxaas-infra.  On GCP using bucket storage for state is probably the nicest approach, i'd certainly prefer that over local state (you need to store it somewhere and can often contain secrets).

Reply:
> I don't think that anyone wanting to run JX in production should be using the installer approach. It's great for demos/trials but unless it can be idempotent and handle updates, I would expect anyone to use it. 

Totally agree BTW, my aim was to highlight that Terraform is a good solution but I want to avoid impacting the barrier to entry to Jenkins X.  We need to keep in mind both folks kicking the tires and enterprise teams.  I was proposing catering for both as David mentioned above the docs IMO should use cloud provider commands, similar to the [harder way](https://github.com/jenkins-x-labs/jenkins-x-installer#harder-way), so that if there's any issues folks can see what's being attempted and learn using docs from their cloud provider.  We absolutely need to provider production level features, I was suggesting we start with the more basic and tackle the more mature setups in a second phase.  But I am coming around given the way the team believes it can not have a big overhead, maintain or improve UX and getting started.

Reply:
@daveconde totally agree on everything you just said. It shouldn't be hard to keep, say, a shell script for gcloud up to date and BDD test it so we know that 'doing it the hard way' without terraform always works.

I'm currently using [these](https://github.com/jenkins-x-labs/jenkins-x-installer/blob/master/create_cluster.sh) [2](https://github.com/jenkins-x-labs/jenkins-x-installer/blob/master/setup_resources.sh) bash scripts in a BDD test for the helmfile stuff right now - so we can easily test the scripts work with jx. The main missing bit is to find a way to extract the actual gcloud commands from those scripts and put them onto a page on the jenkins-x.io website so folks can copy/paste the gcloud commands into a terminal if they want to get started quickly without terraform. We then know the gcloud commands work with jx.